src/helpers.js

```diff
@@ -3313,3 +3380,314 @@ window.unzipItem = async function(itemPath) {
+            $(el_item_name_editor).hide();
+            $(el_item_name_editor).val(html_encode($(el_item).attr('data-name')));
+
+            //show error
+            if(err.message){
+                UIAlert(err.message)
+            }
+        },
+    });
+}
+
+/**
+ * Deletes the given item with path.
+ * 
+ * @param {string} path - path of the item to delete 
+ * @returns {Promise<void>}
+ */
+window.delete_item_with_path = async function(path){
+    try{
+        await puter.fs.delete({
+            paths: path,
+            descendantsOnly: false,
+            recursive: true,
+        });
+    }catch(err){
+        UIAlert(err.responseText);
+    }
+}
+
+window.undo_last_action = async()=>{
+    if (actions_history.length > 0) {
+        const last_action = actions_history.pop();
+
+        // Undo the create file action
+        if (last_action.operation === 'create_file' || last_action.operation === 'create_folder') {
+            const lastCreatedItem = last_action.data;
+            undo_create_file_or_folder(lastCreatedItem); 
+        } else if(last_action.operation === 'rename') {
+            const {options, new_name, old_name, old_path, el_item, el_item_name, el_item_icon, el_item_name_editor, website_url}  = last_action.data;
+            rename_file(options, old_name, new_name, old_path, el_item, el_item_name, el_item_icon, el_item_name_editor, website_url, true); 
+        } else if(last_action.operation === 'upload') {
+            const files = last_action.data;
+            undo_upload(files);
+        } else if(last_action.operation === 'copy') {
+            const files = last_action.data;
+            undo_copy(files);
+        } else if(last_action.operation === 'move') {
+            const items = last_action.data;
+            undo_move(items);
+        } else if(last_action.operation === 'delete') {
+            const items = last_action.data;
+            undo_delete(items);
+        }
+    }
+}
+
+window.undo_create_file_or_folder = async(item)=>{
+    await window.delete_item(item);
+}
+
+window.undo_upload = async(files)=>{
+    for (const file of files) {
+        await window.delete_item_with_path(file);
+    }
+}
+
+window.undo_copy = async(files)=>{
+    for (const file of files) {
+        await window.delete_item_with_path(file);
+    }
+}
+
+window.undo_move = async(items)=>{
+    for (const item of items) {
+        const el = await get_html_from_options(item.options);
+        console.log(item.original_path)
+        move_items([el], path.dirname(item.original_path), true);
+    }
+}
+
+window.undo_delete = async(items)=>{
+    for (const item of items) {
+        const el = await get_html_from_options(item.options);
+        let metadata = $(el).attr('data-metadata') === '' ? {} : JSON.parse($(el).attr('data-metadata'))
+        move_items([el], path.dirname(metadata.original_path), true);
+    }
+}
+
+
+window.get_html_element_from_options = async function(options){
```

Comment:
@jelveh this is the function

Reply:
My bad. I had renamed it at the last moment but forgot to rename the references. I have made a commit to fix it.

Reply:
All fixed now, thank you.